File: classicML/backend/python/callbacks.py
"""classicML的回调函数."""


class History(object):
    """保存训练的历史记录.

    Attributes:
        name: str, default='history',
            历史记录的名称.
        loss_name: str, default='loss'
            使用损失函数的名称.
        metric_name: str, default='metric'
            使用评估函数的名称.
        loss: list, 损失值组成的列表
        metric: list, 评估值组成的列表

    Notes:
        - 使用历史记录数据, 会导致运行速度的降低.
    """
    def __init__(self, name='history', loss_name='loss', metric_name='metric'):
        """
        Arguments:
            name: str, default='history',
                历史记录的名称.
            loss_name: str, default='loss'
                使用损失函数的名称.
            metric_name: str, default='metric'
                使用评估函数的名称.
        """
        self.name = name
        self.loss_name = loss_name
        self.metric_name = metric_name

        self.loss = []
        self.metric = []

    def __call__(self, loss_value, metric_value):
        """记录当前的信息.

        Arguments:
            loss_value: float, 当前的损失值.
            metric_value: float, 当前的评估值.
        """
        self.loss.append(loss_value)
        self.metric.append(metric_value)

File: classicML/backend/cc/callbacks.cc
//
// Created by 孙瑞琦 on 2021/6/11.
//

#include "callbacks.h"

callbacks::History::History() {
    this->name = "history";
    this->loss_name = "loss";
    this->metric_name = "metric";
}

callbacks::History::History(std::string name, std::string loss_name, std::string metric_name) {
    this->name = std::move(name);
    this->loss_name = std::move(loss_name);
    this->metric_name = std::move(metric_name);
}

// TODO(Steve R. Sun, tag:code): backend.cc.callbacks.0.1.2 的测试性能相较Python版本略慢0.3us.
// 记录训练过程中的历史记录, 输入为损失和评估函数值.
void callbacks::History::PyCall(const double &loss_value, const double &metric_value) {
    this->loss.push_back(loss_value);
    this->metric.push_back(metric_value);
}